### ISSUE TITLE
Is the retrieval engine trained one for each dataset?

### ISSUE THREAD

**Issue by liumc14** [meaningful96/CoopRAG#3]
Do you train a separate retrieval engine for each dataset? Why not use several datasets together to train the retrieval engine?
https://github.com/meaningful96/CoopRAG/blob/a2b4282f60b0253688720a83e0b4687ae9a8504e/RaLa/scripts/train_2wiki.sh#L6

**Comment by meaningful96** [comment]
Thanks for the question.

We fine-tune the retriever separately for each dataset to ensure an apples-to-apples comparison with prior work. The baselines we compare against are all trained (or fine-tuned) on a single target dataset and reported that way, so we matched that setup for a fair evaluation.

We did not train a single retriever jointly on multiple datasets in this study—there wasn’t a specific blocker; it simply wasn’t explored for this paper. As a step toward a more general model, we’re currently fine-tuning RaLa on MTEB’s supervised data to report generalized performance, and we plan to release the resulting checkpoint on Hugging Face.

**Comment by liumc14** [comment]
> Thanks for the question.
> 
> We fine-tune the retriever separately for each dataset to ensure an apples-to-apples comparison with prior work. The baselines we compare against are all trained (or fine-tuned) on a single target dataset and reported that way, so we matched that setup for a fair evaluation.
> 
> We did not train a single retriever jointly on multiple datasets in this study—there wasn’t a specific blocker; it simply wasn’t explored for this paper. As a step toward a more general model, we’re currently fine-tuning RaLa on MTEB’s supervised data to report generalized performance, and we plan to release the resulting checkpoint on Hugging Face.

Okay, thank you for your reply. I have another question. Your work solved the multi-hop problem very well using only a single retrieval-generation process and achieved surprising performance. I'm wondering if it can be extended to an iterative retrieval framework? As you mentioned in the appendix of your paper, you combined ircot with cooprag. How is this implemented? Does ircot perform a retrieval using question expansion for each iteration of the generated query? I would greatly appreciate your help in resolving this confusion.

**Comment by meaningful96** [comment]
Yes. In an IRCoT-style multi-step setup, each iteration performs retrieval with an updated unrolled question: you update the unrolled question using evidence extracted from the previous top-k results, then re-run retrieval with that updated unrolled question. The loop is essentially **“unrolled question → retrieve → extract a precise evidence sentence → append the evidence to the unrolled question → retrieve again,”** and only after finishing the loop do you generate the final reasoning and answer.

Concretely, the process looks like this.

On the first step, you construct an unrolled question (e.g., original question + sub-questions + uncertain reasoning cues) and run your retriever (e.g., RaLa) to get top-k documents. Then, instead of feeding all top-k documents to the LLM and immediately answering, you ask the LLM to select the single most directly useful document and extract the key evidence sentence/span as close to verbatim as possible. The point of extracting rather than paraphrasing is to reduce query drift and preserve stable entities/relations needed for the next hop.

Next, you append the extracted evidence sentence to the end of the unrolled question to form an updated unrolled question, and you run retrieval again with this updated unrolled question. You repeat this loop for a fixed number of steps or until a stopping criterion is met (e.g., enough evidence accumulated / top-N evidence set filled).

Finally, you use the accumulated evidence documents/sentences to complete the reasoning chain and produce the final answer. In other words, iterative retrieval here is less about “answering at every step” and more about progressively gathering high-quality evidence that unlocks the next hop.

**Comment by liumc14** [comment]
> Yes. In an IRCoT-style multi-step setup, each iteration performs retrieval with an updated unrolled question: you update the unrolled question using evidence extracted from the previous top-k results, then re-run retrieval with that updated unrolled question. The loop is essentially **“unrolled question → retrieve → extract a precise evidence sentence → append the evidence to the unrolled question → retrieve again,”** and only after finishing the loop do you generate the final reasoning and answer.
> 
> Concretely, the process looks like this.
> 
> On the first step, you construct an unrolled question (e.g., original question + sub-questions + uncertain reasoning cues) and run your retriever (e.g., RaLa) to get top-k documents. Then, instead of feeding all top-k documents to the LLM and immediately answering, you ask the LLM to select the single most directly useful document and extract the key evidence sentence/span as close to verbatim as possible. The point of extracting rather than paraphrasing is to reduce query drift and preserve stable entities/relations needed for the next hop.
> 
> Next, you append the extracted evidence sentence to the end of the unrolled question to form an updated unrolled question, and you run retrieval again with this updated unrolled question. You repeat this loop for a fixed number of steps or until a stopping criterion is met (e.g., enough evidence accumulated / top-N evidence set filled).
> 
> Finally, you use the accumulated evidence documents/sentences to complete the reasoning chain and produce the final answer. In other words, iterative retrieval here is less about “answering at every step” and more about progressively gathering high-quality evidence that unlocks the next hop.

How is the training dataset for the retrieval constructed? Is it constructed using 1000 samples extracted from three multi-hop datasets, as described in the paper, and then expanding the questions so that each expanded question corresponds to multiple positive sample paragraphs, thus generating multiple samples from each expanded question, ultimately forming the 1000 original samples that constitute the training set for the retrieval system? If not, how is it constructed?

**Comment by meaningful96** [comment]
No. The “1,000 samples” mentioned in the paper are not the training set; they are the test (evaluation) set described in Section 4.1. In particular, the paper evaluates on 1,000 test questions from each multi-hop dataset (HotpotQA, MuSiQue, and 2Wiki), and it reuses the same test set used by HippoRAG to ensure a fair comparison against existing baselines. 

The retriever training data is constructed separately and is not formed by “starting from 1,000 samples and expanding them back into 1,000 training samples.” As described in Section 3.6, for retriever fine-tuning the authors first apply question unrolling to questions in the training split to obtain unrolled questions, then pair each unrolled question with one positive document and one hard negative document (following the ColBERT setup) and train the retriever using an InfoNCE-style contrastive objective. 

We also applied the same question-unrolling preprocessing to the training data, and we have already updated the processed version on [HuggingFace: meaningful96/CoopRAG](meaningful96/CoopRAG).

**Comment by liumc14** [comment]
> No. The “1,000 samples” mentioned in the paper are not the training set; they are the test (evaluation) set described in Section 4.1. In particular, the paper evaluates on 1,000 test questions from each multi-hop dataset (HotpotQA, MuSiQue, and 2Wiki), and it reuses the same test set used by HippoRAG to ensure a fair comparison against existing baselines.
> 
> The retriever training data is constructed separately and is not formed by “starting from 1,000 samples and expanding them back into 1,000 training samples.” As described in Section 3.6, for retriever fine-tuning the authors first apply question unrolling to questions in the training split to obtain unrolled questions, then pair each unrolled question with one positive document and one hard negative document (following the ColBERT setup) and train the retriever using an InfoNCE-style contrastive objective.
> 
> We also applied the same question-unrolling preprocessing to the training data, and we have already updated the processed version on [HuggingFace: meaningful96/CoopRAG](meaningful96/CoopRAG).

So, is the publicly available dataset, such as the training set in Hotpot, actually an expansion of the original multi-hop Hotpot dataset's training set? Furthermore, is the test set the same as the 1000 samples evaluated in the experimental section of the paper? Thank you very much!

**Comment by liumc14** [comment]
> No. The “1,000 samples” mentioned in the paper are not the training set; they are the test (evaluation) set described in Section 4.1. In particular, the paper evaluates on 1,000 test questions from each multi-hop dataset (HotpotQA, MuSiQue, and 2Wiki), and it reuses the same test set used by HippoRAG to ensure a fair comparison against existing baselines.
> 
> The retriever training data is constructed separately and is not formed by “starting from 1,000 samples and expanding them back into 1,000 training samples.” As described in Section 3.6, for retriever fine-tuning the authors first apply question unrolling to questions in the training split to obtain unrolled questions, then pair each unrolled question with one positive document and one hard negative document (following the ColBERT setup) and train the retriever using an InfoNCE-style contrastive objective.
> 
> We also applied the same question-unrolling preprocessing to the training data, and we have already updated the processed version on [HuggingFace: meaningful96/CoopRAG](meaningful96/CoopRAG).

I noticed in your paper that the test set follows the method used in Hipporaprag. I also checked the Hipporaprag paper, and it also follows the method used in Ircot to extract the test set, i.e., randomly sampling from the validation set. However, I found that the test set and validation set in your published dataset do not overlap. Why is that? I look forward to your reply. Thank you.
<img width="724" height="256" alt="Image" src="https://github.com/user-attachments/assets/d2e51729-242c-471e-892e-eaa9c53b3288" />

<img width="1104" height="252" alt="Image" src="https://github.com/user-attachments/assets/eedb06c5-2a8c-4761-ab47-8c4fe98c5e85" />

<img width="1078" height="181" alt="Image" src="https://github.com/user-attachments/assets/6e190470-af73-4009-a787-67548c1f6174" />
